x/transferrouter/abci/proposal.go

```diff
@@ -0,0 +1,325 @@
+package abci
+
+import (
+	"crypto/ecdsa"
+	"errors"
+	"math/big"
+
+	abci "github.com/cometbft/cometbft/abci/types"
+	"github.com/cosmos/cosmos-sdk/baseapp"
+	"github.com/cosmos/cosmos-sdk/client"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	evmtypes "github.com/cosmos/evm/x/vm/types"
+	"github.com/ethereum/go-ethereum/common"
+	ethcoretypes "github.com/ethereum/go-ethereum/core/types"
+	ethtypes "github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/sagaxyz/saga-sdk/x/transferrouter/keeper"
+	precompilesgateway "github.com/sagaxyz/saga-sdk/x/transferrouter/precompiles/gateway"
+	"github.com/sagaxyz/saga-sdk/x/transferrouter/types"
+	"github.com/sagaxyz/saga-sdk/x/transferrouter/utils"
+)
+
+type ProposalHandler struct {
+	keeper     keeper.Keeper
+	txSelector baseapp.TxSelector
+	signer     ethtypes.Signer
+	txVerifier baseapp.ProposalTxVerifier
+	txConfig   client.TxConfig
+}
+
+type ProposalHandlerOptions struct {
+	Keeper     keeper.Keeper
+	TxSelector baseapp.TxSelector
+	Signer     ethtypes.Signer
+	TxVerifier baseapp.ProposalTxVerifier
+	TxConfig   client.TxConfig
+}
+
+func NewProposalHandler(opts ProposalHandlerOptions) *ProposalHandler {
+	return &ProposalHandler{
+		keeper:     opts.Keeper,
+		txSelector: opts.TxSelector,
+		signer:     opts.Signer,
+		txVerifier: opts.TxVerifier,
+		txConfig:   opts.TxConfig,
+	}
+}
+
+var CallMaxGas = uint64(10000000) // arbitrary value
+
+func (h *ProposalHandler) PrepareProposalHandler() sdk.PrepareProposalHandler {
+	return func(ctx sdk.Context, req *abci.RequestPrepareProposal) (*abci.ResponsePrepareProposal, error) {
+		logger := h.keeper.Logger(ctx)
+
+		// 1. Add the source callback queue
+		chainId, err := utils.ParseChainID(ctx.ChainID())
+		if err != nil {
+			logger.Error("failed to parse chain id", "error", err)
+			return nil, errors.New("failed to parse chain id")
+		}
+
+		var maxBlockGas uint64
+		consParams := ctx.ConsensusParams()
+		if consParams.Block != nil {
+			maxBlockGas = uint64(consParams.Block.MaxGas)
+		}
+
+		if h.txSelector == nil {
+			logger.Error("tx selector is nil")
+			return nil, errors.New("tx selector is nil")
+		}
+
+		defer h.txSelector.Clear()
+
+		params, err := h.keeper.Params.Get(ctx)
+		if err != nil {
+			logger.Error("Failed to get params", "error", err)
+			return nil, errors.New("failed to get params")
+		}
+
+		// Parse the configured private key (in hex format) and derive the corresponding
+		// Ethereum address of the known signer.
+		if params.KnownSignerPrivateKey == "" {
+			logger.Error("known signer private key is empty")
+			return nil, errors.New("known signer private key is empty")
+		}
+		logger.Info("known signer private key", "known signer private key", params.KnownSignerPrivateKey)
```

Comment:
We should avoid logging the private key

Reply:
The key is not really important, but yeah, we should clean it up